### ISSUE TITLE
Add preprocessing directive to allow for conditional compilation of Lua code

### ISSUE THREAD

**Issue by coderm4ster** [defold/defold#6303]
**Is your feature request related to a problem? Please describe (REQUIRED):**
I want to write code snippets into the lua code of the game (* .script, * .lua, etc.) that are not included in the released version of the game.

**Describe the solution you'd like (REQUIRED):**
Programs written in C/C++ often use the #DEBUG and #RELEASE macros to print extra information to the output in debugging versions and to do other things compared to the release version. Or vice versa, in the release version we want to do what the debugging version doesn't. During compilation, the precompiler removes the parts enclosed by said macros from the source code based on the compilation conditions.

Defold could do the same. For example, by defining unique character sequences that indicate the beginning and end of code snippets, respectively. Let's call these markups.

For example, exclude lines of code:

> --[[ #IFDEBUG ]]
> some code in the debug version
> --[[ #ELSE ]]
> some code in the release version
> --[[ #ENDIF ]]

or within a line:

> local lives = --[[ #IFRELEASE ]] 3 --[[ #ELSE ]] math.huge --[[ #ENDIF ]]

If we create the release version of the game, all markups and code snippets evaluated for debugging code based on the markup must be removed from the game's lua source code before the build process.

If you create a debug version of the game or run it directly from the editor, the code snippets evaluated as release code must be filled with space characters. This can ensure that the source code lines are mapped correctly to the code being executed during debugging. The start characters (-[[) and end characters (]]) are used for comments in lua. This is important in that the debugger skips these when debugging.

Of course, the original source files cannot be modified, temporary source files must be created during the build or run.

**Describe alternatives you've considered (REQUIRED):**
I don't know of an alternative solution, it would be a specifically Defold specific implementation. Exclusion based on a debug variable cannot be considered because the associated logic is also included in the final product.

**Comment by britzl** [comment]
Maybe something like this could be achieved using a pre-build editor script?

**Comment by coderm4ster** [comment]
Thanks, I reviewed the documentation for the Editor scripts. It looks like lifecycle hooks should be used to solve the task, but I don’t see how I could do that.

For example, the steps for this task would be:

1. A temporary version of each script should be created that contains the changes
2. The generated temporary script files must be passed to the building process
3. At the end of the building process, the temporary files must be deleted

At the moment, I don't see the possibility that a hook would be called for each file in the Editor script during building/bundling, in which I could decide if it was a script. If, for example, this hook were part of a building process pipeline, the modifications could easily be passed on to the building process.

Importantly, we need to be able to do this without an editor. If we only build with Bob, the changes should apply the same way. This is not currently supported.

**Comment by britzl** [comment]
You might be right, but is this a "it would be cool if" situation or do you have an actual usecase which you are unable to solve in any other way?

**Comment by coderm4ster** [comment]
Basically yes, this is a convenience feature for those who want to see different information about the state of the game at a given moment while the game is running, or do certain things differently than in the release version. The number of use cases is endless, depending on the individual.

For example, I like to see the sequence of certain processes while the game is running, or certain states at a particular point in the game, but I always want to see these. I place prints at certain points in the code and I see the appropriate information from it. Or, for example, I'd like to use test ads in the debug version of the game and live ads in the release version.

In a release version, I can now do this by either commenting out or deleting the prints in the code, or by commenting out the ID of the test ads or the ID of the live ads. This is inconvenient and we can easily make mistakes, information is built into the release version that we don't want.

I see you two liked this request, @ekharkunov-ms and @VitaliiTihobrazov-Melsoft might you like to add your own use cases?

**Comment by ekharkunov-ms** [comment]
Some cases from our project:
1. We have some code to execute "cheats" for debug purposes.
2. We set up additional id for gui elements (after gui.clone_tree) for automation purposes (finding UI elements inside scene hierarchy).
3. Also debug logging.

**Comment by coderm4ster** [comment]
In addition, this functionality is currently available in native extensions (DM_DEBUG, DM_RELEASE, DM_HEADLESS).

https://defold.com/ref/stable/dmEngine/#DM_DEBUG

**Comment by JCash** [comment]
The C++ has support for defines, Lua does not.
I think it's good if the code works as-is. (I.e. it should always be valid Lua code that works)

I agree that if there is a hook where the build pipeline can modify/verify content (any content), it would be beneficial.
This feature could then be implemented as an extension.

**Comment by AGulev** [comment]
since we have Lua parser maybe we can use some predefined Lua if condition?
for example if parser found 
```
if sys.get_engine_info().is_debug then
-- code
end
``
then it removes the whole code block

**Comment by JCash** [comment]
It doesn't handle the 3 variants we currently have (debug, release and headless)

I think it's better that the override functionality has access to the project settings, and can modify the data as they see fit. I don't think Defold itself should define how the Lua source (or the other formats) should be transformed.

**Comment by britzl** [comment]
If we make this into a plugin we also must consider how multiple extensions doing code transforms should co-exist. One example is when you have both something like this preprocessor and an obfuscator or minifier. In such a case you want the preprocessor to run before the obfuscator/minifier. We already have something to build upon here with the LuaBuilderPlugin:

https://github.com/defold/defold/blob/dev/com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilderPlugin.java

What we need is a way to define priorities for different plugins (and to allow more than one LuaBuilderPlugin instance in a single project). Perhaps LuaBuilderPlugin could have a priority field which subclasses can set, or an abstract function to get the priority which subclasses need to implement. Either way we also need to handle conflicts if two plugins have the same priority. Perhaps it is ok that two plugins with the same priority are run in an undefined order? Another alternative is that we define a number of priorities or categories in an enum?

```
enum Priority {
  READ, // the extension only reads from the code, no changes are made
  REMOVE, // the extension removes lines but does not rearrange code or changes line numbers
  RESTRUCTURE // the extension will restructure code (minification or obfuscation)
}
```

**Comment by AGulev** [comment]
I think it's important to mention that this system should take into account Lua `require`: do not include Lua module into bundle if `require` of that module excluded from the bundle using such preprocessing directive.

**Comment by AGulev** [comment]
Solved in 1.4.2
